README.md
# AuditTool
Web Application for audits, the main function are:
- Track audits activities
- Monitoring progress
- Archiving documentation

This application is implemented with the support of Group Internal Audit. The application is realized with the Meteor-Kitchen framework.


### Quick start with docker-compose
(Prerequisites: docker, docker-compose)

    git clone https://github.com/mttstt/AuditTool.git
    sudo service docker restart
    cd AuditTool
    chmod +x lin-docker.sh
    ./lin-docker.sh -u [passwordAD] [Docker-Hub release] (e.g.: ./lin-docker -u Passw0rd 0.0.16)
    (Jsreport) Access to http://ip-host:5488 and import jsreport.zip contained in git repository
    http://ip-host:81 (AuditTool)
  
  
### For development/test
(Prerequisites: meteor, meteorkitchen, docker)

    sudo docker run -d -p 5488:5488 --restart always jsreport/jsreport:2.0.0-full
    git clone https://github.com/mttstt/AuditTool.git
    cd AuditTool
    ./lin-docker.sh -m
    http://ip-host:3000 (Audittool)
    http://ip-host:5488 (jsreport)


CHANGELOG.md
### 0.0.16 (23-07-2019)
##### Code
* Fixed bug (attach files to the first creation of a test)
##### Enhanced features
* Text colored on monitoring cards
* Modified Dashboard title to Audit Dashboard
* Disabled audit details from Dashboard
* The "Activity type" in Audit Dashboard is hided
##### Enhanced Docker
* Upgrade to (abernix/meteord:node-8.15.1-onbuild)


### 0.0.14 (14-12-2018)
##### Code
* Fixed bug (audit status on tests/update and on issues/update)
* Fixed bug (thumbnail on company field)
* Fixed bug (DD/MM/YYYY stardate in dashboard)


### 0.0.13 (07-12-2018)
##### License change
* License changed. From MIT to GPLv2 because of CKEditor5.
##### New features
* Added icon of audit status on dashboard pages
* Improved audit status bar
##### Code
* Fixed bug (select option of subprocess)
* Upgraded to CKEditor5 and deleted CKEditor4 from the github project
* Deleted CKEditor4 from the github project
* Reordered Github directories
##### Docker
* Upgraded docker (abernix/meteord node-8.14.0-onbuild) 


### 0.0.12 (19-11-2018)
##### Clean code
* Code reviewed and semplified
* Fixed some bugs


### 0.0.11 (09-11-2018)
##### New features
* Add doker functiona in lin.docker.sh (start as daemon, stop)
##### Clean code
* Fixed multiple bugs
* Fixed Jsreport Template


### 0.0.10 (01-11-2018)
##### New features
* Added administrator functions
* DEMO: data now moved on "/private" app directory
#### Clean code
* MK element, alphabetical ordered
* MK elements ("Dest. Selector" & input_template_code) used for the first time
* Added first only collections client, for aggregate results
* Added "ostrio:meteor-root" package
* Added "harrison:papa-parse" package
* Added "themeteorchef:bert" package
* Added "check" package
* Removed "csv-parser" package
 

### 0.0.9 (16-10-2018)
##### New features
* Added Monitoring detection Controls section (registry, detection and publising)
#### Clean code
* Added "jcbernack:reactive-aggregate" package


### 0.0.8 (28-09-2018)
##### Clean code
* Removed "mikowals:batch-insert" package
* Removed "dburles:mongo-collection-instances" package
* Added "perak:c3" package
* Added "random" package
* Reviewed Access rights of collection Activies 
* Integration with MK (delete some import from externals file)
* Integration with Meteor users and added the new group: Auditors
* Upgraded of jsreport
##### New features
* Traslation in english
* Introduction Graph for continuos controls


### 0.0.7 (27-02-2018)
##### New features
* New architecture on docker
* Created new script lin-docker.sh in order to manage 3 containers: jsreport, mongodb, audittool
##### Clean code
* Renamed some files


### 0.0.6 (16-12-2017)
##### New features
*	Used the car.sci numbering for the checks.
*	Auditing numbering with the change of status "approved"
* Updated of the verbal template pdf
##### BUG
*	Automatic audit-number
##### Clean code
* Eliminated copy of an external file (Bootstrap plugin-file)
* Used new features of MK release (0.9.90):	Eliminate imports of some external Files
* Updated file batch go-windows.bat
* Upgraded attachment-box
##### Added MIT License


### 0.0.5 (06-08-2017)
##### New features
* Improved "status" management


### 0.0.4 (15-07-2017)
##### New Features
* JSReport implementation (NPM jsreport-browser-client-dist) for pdf file


### 0.0.3 (15-06-2017)
##### New Features
* PDF: Introduced skip Init phase creation new element 
* Added Library Element


### 0.0.2 (15-05-2017)
##### New Features
* Added Attachments Element (raix:ui-dropped-event)


### 0.0.1 (15-04-2017)
##### New Features
* Added advanced editor CKEDITOR
* Noe read file from files on change (mikowals:batch-insert, dburles:mongo-collection-instances, NPM csv-parser)


### 0.0.0 (15-03-2017)
##### New Features
* First release with meteor-kitchen


insert.js
var pageSession = new ReactiveDict();

Template.HomePrivateInsert.onCreated(function() {
	
});

Template.HomePrivateInsert.onDestroyed(function() {
	
});

Template.HomePrivateInsert.onRendered(function() {
	
	Meteor.defer(function() {
		globalOnRendered();
		$("input[autofocus]").focus();
	});
});

Template.HomePrivateInsert.events({
	
});

Template.HomePrivateInsert.helpers({
	
});

Template.HomePrivateInsertInsertAudit.onCreated(function() {
	
});

Template.HomePrivateInsertInsertAudit.onDestroyed(function() {
	
});

Template.HomePrivateInsertInsertAudit.onRendered(function() {
	

	pageSession.set("homePrivateInsertInsertAuditInfoMessage", "");
	pageSession.set("homePrivateInsertInsertAuditErrorMessage", "");

	$(".input-group.date").each(function() {
		var format = $(this).find("input[type='text']").attr("data-format");

		if(format) {
			format = format.toLowerCase();
		}
		else {
			format = "mm/dd/yyyy";
		}

		$(this).datepicker({
			autoclose: true,
			todayHighlight: true,
			todayBtn: true,
			forceParse: false,
			keyboardNavigation: false,
			format: format
		});
	});

	$("input[type='file']").fileinput();
	$("select[data-role='tagsinput']").tagsinput();
	$(".bootstrap-tagsinput").addClass("form-control");
	$("input[autofocus]").focus();
});

Template.HomePrivateInsertInsertAudit.events({
	"submit": function(e, t) {
		e.preventDefault();
      
		pageSession.set("homePrivateInsertInsertAuditInfoMessage", "");
		pageSession.set("homePrivateInsertInsertAuditErrorMessage", "");

		var self = this;

		function submitAction(result, msg) {
			var homePrivateInsertInsertAuditMode = "insert";
			if(!t.find("#form-cancel-button")) {
				switch(homePrivateInsertInsertAuditMode) {
					case "insert": {
						$(e.target)[0].reset();
					}; break;

					case "update": {
						var message = msg || "Saved.";
						pageSession.set("homePrivateInsertInsertAuditInfoMessage", message);
					}; break;
				}
			}
          
          
          	console.log('nuova attività',result);
          	
            var current_tipo= Activities.findOne({_id:result}).tipo;                             
              
            switch(Activities.findOne({_id:result}).tipo) { 	
              						case "Audit": Router.go("audits.workflow", mergeObjects(Router.currentRouteParams(), {})); break; 
              					  	case "Informativa": Router.go("informativa.workflow", mergeObjects(Router.currentRouteParams(), {})); break; 
              						case "Audit Filiale": Router.go("audit_filiale.workflow", mergeObjects(Router.currentRouteParams(), {})); break; }
			}

		function errorAction(msg) {
			msg = msg || "";
			var message = msg.message || msg || "Error.";
			pageSession.set("homePrivateInsertInsertAuditErrorMessage", message);
		}

		validateForm(
			$(e.target),
			function(fieldName, fieldValue) {

			},
			function(msg) {

			},
			function(values) {
             	                values.data = moment().format('MM/DD/YYYY');
			console.log(values.data);
                                values.auditors = Meteor.user();
			console.log('a',values.auditors);
                                values.status = "Avviato";                               	

                    Meteor.call("activitiesInsert", values, function(e, r) { if(e) errorAction(e); else submitAction(r); });
			}
		);

		return false;
	},
	"click #form-cancel-button": function(e, t) {
		e.preventDefault();

		

		/*CANCEL_REDIRECT*/
	},
	"click #form-close-button": function(e, t) {
		e.preventDefault();

		/*CLOSE_REDIRECT*/
	},
	"click #form-back-button": function(e, t) {
		e.preventDefault();

		/*BACK_REDIRECT*/
	}

	
});

Template.HomePrivateInsertInsertAudit.helpers({
	"infoMessage": function() {
		return pageSession.get("homePrivateInsertInsertAuditInfoMessage");
	},
	"errorMessage": function() {
		return pageSession.get("homePrivateInsertInsertAuditErrorMessage");
	}
	
});


lin-docker2.sh
#!/bin/bash
#
# -n: cancella tutto (container, images,..) e ricrea la struttura 
#
# newgrp docker
#

while [[ $# -gt 0 ]]
do
key="$1"

case $key in
    -n|--new)
    echo "Clean ALL"

        echo "Ferma tutti i containers"
        docker container stop $(docker ps -a -q)

        echo "Elimina tutti i container"
        docker container rm $(docker ps -a -q)

        echo "Elimina tutti i volumi"
        docker volume rm $(docker volume ls -q)

        echo "Crea volumi"
        docker volume create audittoolvolume
        docker volume create jsreportvolume
        docker volume create mongovolume

        # Per copiare i dati devo prima creare un container dummy temporaneo
        echo "Copio file lib"
        docker container create --name dummy -v audittoolvolume:/tmp/files/lib hello-world
        docker cp ~/AuditTool/files/lib dummy:/tmp/files
        docker rm dummy

        echo "Elimina tutte le immagini"
        docker rmi $(docker images -a -q)

        echo "Elimina tutte le reti tranne quella di default 172.17.0.0/16"
        docker network rm mttlan

        echo "Crea nuova sottorete mttlan"
        docker network create --subnet=192.168.2.0/16 mttlan
       
        # Avvia container
        docker run -d --net mttlan --ip 192.168.2.1 --name jsreport -p 80:5488 --restart always -v jsreportvolume:/jsreport jsreport/jsreport
        docker run -d --net mttlan --ip 192.168.2.2 --name meteor-mongo -v mongovolume:/data/db mongo

    shift # past argument
    shift # past value
    ;;
    -b|--bye)
    echo "bye"

    echo "bye"
    shift # past argument
    shift # past argument
    ;;
    *)    # unknown option
    shift # past argument
    ;;
esac
done

echo "Inizio"

# sudo systemctl stop firewalld.service
# echo "In order to list images: docker images -a"
# echo "In order to delete container: docker rm <ID container>"
# echo "In order to delete: docker rmi <ID Images>"
# echo "In order to delete all: docker system prune"
# echo "In order to install Mongodb: docker run --name mongo-meteor -d mongo"
# echo "In order to create network: docker network create --subnet=192.168.1.0/16 mttlan"

rm -rf .git

cd ~/AuditTool

cp -R files/ /tmp

sudo rm -fR AuditTool

wget http://www.meteorkitchen.com/api/getapp/json/Tqq4JcxsuGEBZrben -O AuditTool.json

meteor-kitchen AuditTool.json AuditTool

cd  AuditTool

meteor build --architecture=os.linux.x86_64 ./

############################################################## for docker ##########################################################
# Se attivo ferma ed elimina il container audittool  
docker stop audittool || true && docker rm audittool || true

docker run \
 -it \
 --net mttlan \
 --ip 192.168.2.3 \
 --name audittool \
 -e MONGO_URL=mongodb://192.168.2.2 \
 -e ROOT_URL=http://192.168.2.3 \
 -v audittoolvolume:/tmp/files/lib \
 -v /home/mtt/AuditTool/AuditTool:/bundle \
 -p 8080:80 \
 abernix/meteord:node-8.9.3-base \
 /bin/bash

echo "Docker images ls:"
docker image ls

docker ps -q | xargs docker inspect --format '{{ .Id }} - {{ .Name }} - {{ .NetworkSettings.IPAddress }}'
################################################################################################################################


go-linux-docker.sh
#!/bin/bash
jsreport init &
jsreport start &
# sudo systemctl stop firewalld.service
echo "In order to list images: docker images -a"
echo "In order to delete container: docker rm <ID container>"
echo "In order to delete: docker rmi <ID Images>"
echo "In order to delete all: docker system prune"
echo "In order to install Mongodb: docker run --name mongo-meteor -d mongo"
echo "In order to create network: docker network create --subnet=192.168.1.0/16 mttlan"

cd /home/mtt/AuditTool

cp -R files/ /tmp

rm -fR AuditTool

wget http://www.meteorkitchen.com/api/getapp/json/Tqq4JcxsuGEBZrben -O AuditTool.json

meteor-kitchen AuditTool.json AuditTool

meteor create AuditTool

cd  AuditTool

############# for docker ##########################################################
docker system prune -f
docker network create --subnet=192.168.1.0/16 mttlan
docker run --name mongo-meteor -d mongo
echo "FROM jshimko/meteor-launchpad:latest" > ~/AuditTool/AuditTool/Dockerfile
cp ~/AuditTool/files/.dockerignore  ~/AuditTool/AuditTool/.dockerignore
cp ~/AuditTool/files/docker-compose.yml ~/AuditTool/AuditTool/docker-compose.yml
sudo docker build -t audittool .
echo "docker images ls"
docker image ls
docker run -d -e MONGO_URL=mongodb://172.17.0.2 -P audittool
###################################################################################

# docker run -d --name audittool -P audittool
docker ps -q | xargs docker inspect --format '{{ .Id }} - {{ .Name }} - {{ .NetworkSettings.IPAddress }}'
# docker ps -q | xargs docker inspect --format '{{ .NetworkSettings.IPAddress }}'

#awk 'NR==8{print "password: " "\x27" "Gennaio:2018" "\x27"}7' /home/mtt/AuditTool/AuditTool/server/methods/my-methods.js >  /home/mtt/AuditTool/AuditTool/server/methods/my-methods.js


files/ckeditor/config.js
/**
 * @license Copyright (c) 2003-2017, CKSource - Frederico Knabben. All rights reserved.
 * For licensing, see LICENSE.md or http://ckeditor.com/license
 */
CKEDITOR.editorConfig = function( config ) {
	config.toolbarGroups = [
		{ name: 'document', groups: [ 'mode', 'document', 'doctools' ] },
		{ name: 'clipboard', groups: [ 'clipboard', 'undo' ] },
		{ name: 'editing', groups: [ 'find', 'selection', 'spellchecker', 'editing' ] },
		{ name: 'styles', groups: [ 'styles' ] },
		{ name: 'forms', groups: [ 'forms' ] },
		{ name: 'colors', groups: [ 'colors' ] },
		{ name: 'tools', groups: [ 'tools' ] },
		'/',
		{ name: 'basicstyles', groups: [ 'basicstyles', 'cleanup' ] },
		{ name: 'paragraph', groups: [ 'list', 'indent', 'blocks', 'align', 'bidi', 'paragraph' ] },
		{ name: 'links', groups: [ 'links' ] },
		{ name: 'insert', groups: [ 'insert' ] },
		{ name: 'others', groups: [ 'others' ] },
		{ name: 'about', groups: [ 'about' ] }
	];

	config.removeButtons = 'About,ShowBlocks,Iframe,Source,Form,Checkbox,Radio,TextField,Textarea,Select,Button,ImageButton,HiddenField,Link,Unlink,Anchor,Save,SelectAll,RemoveFormat,CreateDiv,Font,Language,Print,NewPage,Flash,Smiley,PageBreak';
};

lin-docker.sh
#!/bin/bash
#
# Use: lin-docker.sh [command] [Password] [Docker-Hub release] 
# Put AuditTool in a Docker architecture
#
#./lin-docker.sh -u [Password Active Directory] [Docker-Hub release]
#./lin-dcoker.sh -b [Docker-Hub release] 
#./lin-dcoker.sh -p [Password Docker-Hub]
#./lin-dcoker.sh -m [Password Active Directory]
#
# Possible commands:
#
# -h, --help        Help
# -m, --meteor      Launch meteor, without Docker, for testing [Password Active Directory]
# -b, --build       Build audittool docker image [New Docker-Hub release]
# -p, --push        Push audittool image to Docker Hub [Password Docker Hub] [Docker release] 
# -u, --dockerup    Docker-compose up [Password Active Directory] [Docker release]  
# -d, --delete      Delete all (containers, images, volumes, networks)
# -s, --stop	    docker-compose stop, useful when containers are started like service/ademon
# 
# Useful command for ubuntu 17: newgrp docker
#
# Useful command for bash to containet:  docker exec -ti <container> bash

while [[ $# -gt 0 ]]
do
key="$1"
passwd="$2"
release="$3"
statusJreport=`curl -s --head http://localhost:5488 | head -n 1 | grep "HTTP/1.[01] [23].."`

case $key in
    -d|--delete)
        echo "Delete All"
        echo "Ferma tutti i containers"
        docker-compose down
        echo "Elimina tutti i containers"
        docker rm $(docker container ls -a -q) -f
        echo "Elimina tutti i volumi"
        docker volume prune -f
        echo "Elimina tutte le immagini"
        docker rmi $(docker images -a -q)
    shift # past argument
    shift # past argument
    ;;
    
    -b|--build)
        echo "build"
#       cd ~/AuditTool
        if [ ! -f ~/AuditTool/.gitignore ]; then
            echo ".gitignore not found!"
        else
            rm .gitignore
	fi
        rm -rf .git
        rm -fR AuditTool
        wget http://www.meteorkitchen.com/api/getapp/json/Tqq4JcxsuGEBZrben -O AuditTool.json
        meteor-kitchen AuditTool.json AuditTool
        cd AuditTool       
	cp ../docker/Dockerfile .
	cp ../docker/.dockerignore .

	export TAG=$passwd && docker-compose -f ../docker/docker-compose-dev.yml build --force-rm --no-cache 
    shift # past argument
    shift # past argument
    exit 1
    ;;

    -m|--meteor)
        echo "meteor"
#	cd ~/AuditTool
	if [ -z "$statusJreport" ]; then
                echo 'Jsreport not installed: no pdf reports !' >&2
        else
                echo 'Jsreport found !'
        fi

        if [ ! -f ~/AuditTool/.gitignore ]; then
                echo ".gitignore not found!"
        else
            rm .gitignore
        fi
        rm -rf .git     
        rm -fR AuditTool
        wget http://www.meteorkitchen.com/api/getapp/json/Tqq4JcxsuGEBZrben -O AuditTool.json
        meteor-kitchen AuditTool.json AuditTool
        cd AuditTool
        meteor
    shift # past argument
    shift # past argument
    ;;
    
    -p|--push)
        echo "Push image to Docker Hub"
        docker login -password $passwd -username mttstt        
        docker push mttstt/audittool
    shift # past argument
    shift # past argument
    ;;  

    -u|--dockerup)
        echo "Docker up like a service (-d)" 
        export TAG=$release && export passwdAD=$passwd && docker-compose -f docker/docker-compose.yml up -d
	exit 1
    shift # past argument
    shift # past argument
    ;;   

    -s|--stop)
        echo "Stopping AuditTool container daemon"
        docker-compose -f docker/docker-compose.yml stop
    shift # past argument
    shift # past argument
    ;;

    -b|--bye)
    echo "bye"
    echo "bye"
    shift # past argument
    shift # past argument
    ;;   
    
    *)    # unknown option
   
        echo "Use: lin-docker.sh [command] [Password] [Docker HUB release]" 
        echo "Put AuditTool in a Docker architecture"
        echo ""       
        echo "./lin-docker.sh -u [Password Active Directory] [Docker-Hub release]"
        echo "./lin-dcoker.sh -b [New Docker-Hub release]"
        echo "./lin-dcoker.sh -p [Password Docker-Hub]"
        echo ""
        echo "Possible commands:"
        echo ""
        echo "-h, --help        Help"
        echo "-m, --meteor      Launch meteor, without Docker (for testing) [Password Active Directory]"
        echo "-b, --build       Build audittool docker image [New Docker-Hub release]"
        echo "-p, --push        Push audittool image to Docker Hub [Password Docker Hub] [Docker-Hub release]"
        echo "-u, --dockerup    Docker-compose up [Password Active Directory] [DockerHub release]"
        echo "-d, --delete      Delete all (containers, images, volumes, networks)"
        echo "-s, --stop        Stop Audittol container daemon"
      
    shift # past argument# Options:
    ;;
        
esac
done
